KAFKA-15445: Add JVM Docker image

VedarthConfluent opened the pull request:
This PR aims to add Apache Kafka JVM Docker image as per the following KIP - https://cwiki.apache.org/confluence/display/KAFKA/KIP-975%3A+Docker+Image+for+Apache+Kafka

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

VedarthConfluent committed:
Add JVM Docker image

VedarthConfluent committed:
Refactor build test and setup proper directory structure

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Refactor scripts running in Dockerfile

VedarthConfluent committed:
Refactor bash scripts

VedarthConfluent committed:
Remove redundant files

VedarthConfluent committed:
Add licence to the files

ijuma commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,78 @@
+RUN go test ./...
+
+
+FROM eclipse-temurin:17-jre
```
I suggest we go with 21 from the start.

VedarthConfluent replied:
We were waiting for [this](https://hub.docker.com/layers/library/eclipse-temurin/21-jre/images/sha256-0e3f813fbdcd0b62e58c3ad984da29d96e0b78cede4578d2d1d0f8391f1f136c?context=explore) to release. Will be updating it post testing

VedarthConfluent committed:
Updated to 21-jre and fix typos

VedarthConfluent committed:
Add a release script for pushing docker images to dockerhub

VedarthConfluent committed:
Update release script to support registry and add error handling

sanjay-awatramani commented on docker/docker_release.py:
```diff
@@ -0,0 +1,80 @@
+def remove_builder():
+    subprocess.run(["docker", "buildx", "rm", "kafka-builder"])
+
+if __name__ == "__main__":
```
This is an interactive script. How would we support automation of release process via GH actions or Jenkins?

VedarthConfluent replied:
Since the existing release process is driven by an interactive script, we are inclined towards keeping it that way. We will explore and check the feasibility of automating this

VedarthConfluent replied:
We have decided to keep the release and promotion process to be driven by local scripts for now

VedarthConfluent committed:
Remove confluent artifacts from sanity test

Remove confluent kafka python client as dependency
Remove schema registry and connect from compose files
Download kafka binary and use it's scripts to run tests

VedarthConfluent committed:
Remove redundant files

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

ijuma commented on docker/resources/common-scripts/kafka-propertiesSpec.json:
```diff
@@ -0,0 +1,25 @@
+    "CONFLUENT": true
+  },
+  "renamed": {
+    "KAFKA_ZOOKEEPER_CLIENT_CNXN_SOCKET": "zookeeper.clientCnxnSocket"
```
I think we should skip zk since it's deprecated and will be removed in 4.0.

VedarthConfluent replied:
Sounds good! We won't support Zookeeper in the docker image. I have updated the PR to reflect the same

VedarthConfluent committed:
Fix tests

Add SSL_TOPIC again in constants.py

VedarthConfluent committed:
Use CDS to start kafka

sanjay-awatramani commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,101 @@
+    mkdir -p /var/lib/kafka/data /etc/kafka/secrets /var/log/kafka /var/lib/zookeeper; \
+    mkdir -p /etc/kafka/docker /usr/logs; \
+    adduser -h /home/appuser -D --shell /bin/bash appuser; \
+    chown appuser:appuser -R /etc/kafka/ /usr/logs /opt/kafka; \
+    chown appuser:root -R /etc/kafka /var/lib/kafka /etc/kafka/secrets /var/lib/kafka /etc/kafka /var/log/kafka /var/lib/zookeeper; \
```
`/etc/kafka` appears thrice in this segment. The first two occurrences seem redundant as the third one will override the first two `chown`s

VedarthConfluent replied:
Valid point! I have removed the redundant occurrences

sanjay-awatramani commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,101 @@
+    chown appuser:root -R /etc/kafka /var/lib/kafka /etc/kafka/secrets /var/lib/kafka /etc/kafka /var/log/kafka /var/lib/zookeeper; \
+    chmod -R ug+w /etc/kafka /var/lib/kafka /var/lib/kafka /etc/kafka/secrets /etc/kafka /var/log/kafka /var/lib/zookeeper; \
+    rm kafka.tgz; \
+    apk cache clean; \
```
Should we also remove the packages installed by
```apk add --no-cache curl wget gpg dirmngr gpg-agent```
Should we also remove the KEYS added by
```wget -nv -O KEYS https://downloads.apache.org/kafka/KEYS```

VedarthConfluent replied:
This has been incorporated in the PR

sanjay-awatramani commented on docker/resources/common-scripts/configure:
```diff
@@ -0,0 +1,147 @@
+  ub ensure KAFKA_ADVERTISED_LISTENERS
+fi
+
+# By default, LISTENERS is derived from ADVERTISED_LISTENERS by replacing
```
As per [docs](https://kafka.apache.org/documentation/#brokerconfigs_advertised.listeners), ADVERTISED_LISTENERS is derived from LISTENERS so this seems counter intuitive for someone used to Kafka's configs. 
Also ADVERTISED_LISTENERS is optional and we are making it mandatory in above if block
```ub ensure KAFKA_ADVERTISED_LISTENERS```

VedarthConfluent replied:
In case of docker image usage will mostly be external to the docker network, hence it makes sense to keep ADVERTISED_LISTENERS as mandatory. 
Although it might be a bit counter intuitive, but it makes sense in the context of docker image to define ADVERTISED_LISTENERS and derive LISTENERS because of the same reason.

VedarthConfluent committed:
Resolve PR comments

- Remove zookeeper support & tests
- Optimise image by removing redundant packages
- Minor code fixes

sanjay-awatramani commented on docker/jvm/jsa_launch:
```diff
@@ -0,0 +1,15 @@
+PIDS=$!
+
+sleep 10
+echo "test" | opt/kafka/bin/kafka-console-producer.sh --topic test-topic --bootstrap-server localhost:9092
```
- If there is an error at any step while creating JSA (e.g. unable to produce/consume) the JSA would have less than optimal information. We should bubble up the failure by failing the build so that this is identified rather than being suppressed.
- I get a WARN that test-topic doesn't exist which seems that something has gone wrong in the build. Kafka automatically creates a topic so this warning can be ignored, but would be good to create a topic before producing to it so that we don't get unnecessary WARNings in the build output.

VedarthConfluent replied:
- Valid point. Made the changes to reflect the same
- Creating the topic explicitly now

sanjay-awatramani commented on docker/jvm/jsa_launch:
```diff
@@ -0,0 +1,15 @@
+KAFKA_JVM_PERFORMANCE_OPTS="-XX:ArchiveClassesAtExit=kafka.jsa" opt/kafka/bin/kafka-server-start.sh opt/kafka/config/kraft/server.properties &
+PIDS=$!
+
+sleep 10
```
`sleep` is not the best way to wait for an application to be ready. We could check for something that indicates the app is ready (for e.g. checking if 9092 is ready to accept connections) and then maybe sleep a few seconds if required.

VedarthConfluent replied:
Replaced static sleep with conditional loops with timeouts

sanjay-awatramani commented on docker/test/docker_sanity_test.py:
```diff
@@ -0,0 +1,193 @@
+            f.write(s)
+
+        subprocess.run(["docker-compose", "-f", filename, "up", "-d"])
+        time.sleep(25)
```
`sleep` is not the best way to wait for an application to be ready. We could check for something that indicates the app is ready (for e.g. checking if 9092 is ready to accept connections) and then maybe sleep a few seconds if required.

VedarthConfluent replied:
Removed static sleep and relying on client scripts to wait for server, with timeouts

VedarthConfluent committed:
Add github actions workflow for build and test of jvm docker image

VedarthConfluent committed:
Add description to kafka url link

VedarthConfluent committed:
Refactors jsa launch script with error handling and timeouts

VedarthConfluent committed:
Remove static sleep from sanity tests

VedarthConfluent committed:
Rely on scripts to detect when server is up

VedarthConfluent committed:
Removed redundant wait in jsa generation

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Add promotion script

VedarthConfluent committed:
Add requirements.txt for promotion script

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Fix property file location

VedarthConfluent committed:
Add support for supplying properties through file mounting

Changes to add support for file mounting of properties.
Env variables, if defined, will override the file input properties.

VedarthConfluent committed:
Ensure that environment variable configs are always appended in newline

VedarthConfluent committed:
Add missing brackets in template

VedarthConfluent committed:
Add test for file input

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Bubble up test errors to root build test script

VedarthConfluent committed:
Add license comment and refactor ub scripts to remove redundant code

VedarthConfluent committed:
Add readme file and refactor the python scripts

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Add local setup section in Readme

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

VedarthConfluent committed:
Update file input test and readme file

VedarthConfluent committed:
Add test for broker metrics

VedarthConfluent committed:
Remove potential flakiness from the test

nkonev commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,97 @@
+    mkdir -p /etc/kafka/docker /usr/logs /mnt/shared/config; \
+    adduser -h /home/appuser -D --shell /bin/bash appuser; \
+    chown appuser:appuser -R /usr/logs /opt/kafka /mnt/shared/config; \
+    chown appuser:root -R /var/lib/kafka /etc/kafka/secrets /var/lib/kafka /etc/kafka /var/log/kafka; \
```
Duplicated `/var/lib/kafka`

VedarthConfluent replied:
Thanks for catching this. It has been fixed

VedarthConfluent committed:
Remove redundant paths in chown and chmod

VedarthConfluent committed:
Fix build by adding license and excluding files where it cannot be added

VedarthConfluent committed:
Update documentation to include docker image in the release process

omkreddy commented on docker/test/fixtures/secrets/client-ssl.properties:
```diff
@@ -0,0 +1,24 @@
+ssl.keystore.location=./test/fixtures/secrets/client.keystore.jks
+ssl.keystore.password=abcdefgh
+ssl.key.password=abcdefgh
+ssl.enabled.protocols=TLSv1.2,TLSv1.1,TLSv1
```
we can remove `ssl.enabled.protocols`. default should be fine

VedarthConfluent replied:
Removed

VedarthConfluent committed:
Add CDS for storage format

VedarthConfluent commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,98 @@
+    wget -nv -O kafka.tgz "$kafka_url"; \
+    tar xfz kafka.tgz -C /opt/kafka --strip-components 1;
+
+RUN /etc/kafka/docker/jsa_launch
```
Add a comment to describe CDS

VedarthConfluent commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,98 @@
+      org.label-schema.build-date="${build_date}" \
+      org.label-schema.vcs-url="https://github.com/apache/kafka" \
+      org.label-schema.schema-version="1.0" \
+      maintainer="apache"
```
Should be Apache Kafka

VedarthConfluent committed:
Resolve comments

- Update label values and remove redundand packages
- Allow build and test script to be executable from anywhere

VedarthConfluent committed:
Refactor release script

VedarthConfluent committed:
Make RC release script non interactive

VedarthConfluent committed:
Remove the confirmation check from RC release script

VedarthConfluent committed:
Fix test report path in workflow and fix inline docs

VedarthConfluent committed:
Handle error to ensure files get deleted while running tests

VedarthConfluent committed:
Update readme file to reflect the change

VedarthConfluent committed:
Remove redundant configs from tests

VedarthConfluent committed:
Remove advertised.listeners from server properties in controller mode

VedarthConfluent committed:
Copy property files instead of move to fix errors on restart

VedarthConfluent committed:
Remove redundant checks and improve script documentation

VedarthConfluent committed:
Add test for isolated mode kafka cluster

VedarthConfluent committed:
Fix directory naming

VedarthConfluent committed:
Rename ub to utility and fix test compose files

VedarthConfluent committed:
Fix build error

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

omkreddy commented on docker/README.md:
```diff
@@ -0,0 +1,72 @@
+
+Bulding image and running tests using github actions
+----------------------------------------------------
+This is the recommended way to build, test and get a CVE report for the docker image.
```
Can we run github workflows from a branch also? This is because we need to run this workflow from release branch.

VedarthConfluent replied:
Yes, github actions UI provides the option to select the branch that you want to run the workflow on. Note that the kafka containerised will still be fetched from the tarball url. The code to build the container will depend on the branch selected.

omkreddy commented on docker/README.md:
```diff
@@ -0,0 +1,72 @@
+-------------------------------------
+- The image uses the kafka downloaded from provided kafka url
+- The image can be run in a container in default mode by running
+`docker run <image-name:tag> -p 9092:9092`
```
I am getting this error with -p flag

 ```
 docker run kafka-master:latest -p 9092:9092
/__cacert_entrypoint.sh: line 30: exec: -p: invalid option
exec: usage: exec [-cl] [-a name] [command [argument ...]] [redirection ...]
```

VedarthConfluent replied:
Ah thanks for catching this. It's indeed incorrect. I have fixed it in the last commit. Please try `docker run -p 9092:9092 kafka-master:latest`

omkreddy commented on docker/README.md:
```diff
@@ -0,0 +1,72 @@
+            - For abc.def, use KAFKA_ABC_DEF
+            - For abc-def, use KAFKA_ABC___DEF
+            - For abc_def, use KAFKA_ABC__DEF
+- Hence order of precedence of properties is the follwing:-
```
follwing => following

VedarthConfluent committed:
Fix typo errors and remove redundant client config

ashwinpankaj commented on docker/jvm/jsa_launch:
```diff
@@ -0,0 +1,48 @@
+
+KAFKA_JVM_PERFORMANCE_OPTS="-XX:ArchiveClassesAtExit=storage.jsa" opt/kafka/bin/kafka-storage.sh format -t $KAFKA_CLUSTER_ID -c opt/kafka/config/kraft/server.properties
+
+KAFKA_JVM_PERFORMANCE_OPTS="-XX:ArchiveClassesAtExit=kafka.jsa" opt/kafka/bin/kafka-server-start.sh opt/kafka/config/kraft/server.properties &
```
doesn't this overwrite value of KAFKA_JVM_PERFORMANCE_OPTS set in L19 ?

VedarthConfluent replied:
We intend to use this script to generate two separate CDS files for kafka server start and storage format commands. Hence we are intentionally providing different values for `KAFKA_JVM_PERFORMANCE_OPTS` to ensure creation of storage.jsa and kafka.jsa.

ashwinpankaj commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,100 @@
+COPY --chown=appuser:appuser resources/utility/ ./
+
+# Generate utility executable for dealing with env variables
+RUN go build -ldflags="-w -s" ./utility.go
```
this is the first use of golang in this project - can't this utility be implemented in python?

VedarthConfluent replied:
Golang allows us to keep the size of the final image small, as the executable it provides doesn't require golang installation in final image. Keeping the utility in golang allows us to keep the image size small and code well structured and tested.

ashwinpankaj commented on docker/jvm/Dockerfile:
```diff
@@ -0,0 +1,100 @@
+    apk upgrade ; \
+    apk add --no-cache wget gcompat procps netcat-openbsd; \
+    mkdir opt/kafka; \
+    wget -nv -O kafka.tgz "$kafka_url"; \
```
could we add an gpg check ?

VedarthConfluent replied:
Done

ashwinpankaj commented on docker/resources/common-scripts/server.properties:
```diff
@@ -0,0 +1,28 @@
+group.initial.rebalance.delay.ms=0
+inter.broker.listener.name=PLAINTEXT
+listener.security.protocol.map=CONTROLLER:PLAINTEXT,PLAINTEXT:PLAINTEXT,PLAINTEXT_HOST:PLAINTEXT
+listeners=PLAINTEXT://localhost:29092,CONTROLLER://localhost:29093,PLAINTEXT_HOST://0.0.0.0:9092
```
just curious, any reason why the port numbering scheme is different ?

VedarthConfluent replied:
We have removed these configs. We will now be using the default kraft configs provided in kafka (configs/kraft/server.properties) bundled in the kafka tarball

ashwinpankaj commented on docker/README.md:
```diff
@@ -0,0 +1,72 @@
+Bulding image and running tests using github actions
+----------------------------------------------------
+This is the recommended way to build, test and get a CVE report for the docker image.
+Just choose the image type and provide kafka url to `Docker build test` workflow. It will generate a test report and CVE report that can be shared to the community.
```
can we provide examples for image-type and kafka-url please ?

VedarthConfluent replied:
Added description for both of them

VedarthConfluent committed:
Resolve comments

- Add gpg verification in jsa file generation
- Add details in README for image type and kafka url

VedarthConfluent committed:
Update config handling logic

- Remove all default property files from docker folder
- Use default files from kafka binary
- Do not use default kafka configs if any user provided config is found

VedarthConfluent committed:
Make python scripts executable and add detailed examples in README

omkreddy reviewed (approved):
@VedarthConfluent Thanks for the PR. LGTM. I will wait for any other comments before merging the PR.

VedarthConfluent committed:
Remove redundant fix for controller flow

VedarthConfluent committed:
Update readme docs and minor change in github actions workflow

VedarthConfluent committed:
Add empty line at end of all files

VedarthConfluent committed:
Update isolated test compose file

VedarthConfluent committed:
Merge branch 'trunk' of github.com:apache/kafka into docker-image

mimaison commented:
This adds a small piece of code in Go. I think this should have been mentioned in the KIP as adding a new language is something worth discussing. 

I understand Go is well suited for this use case but it would have been nice to compare with Java. For example, if we're only shaving a few milliseconds I'm not sure introducing a new language in the repository is necessarily worth it. We're also looking at having a native image, in that case that logic in Java _could_ match Go.

VedarthConfluent commented:
Hey @mimaison,

Thanks for the input. We will update the KIP based on conclusion of this conversation.

Indeed golang is a new language for the repository and it's a relatively small piece of logic. But in my opinion given the benefit we are getting from using golang, it's worth the maintenance cost of the new language in the repo.

Here are some of our reasons behind the decision to use **golang**:-

- It allows us to have a very lightweight executable that can be used without need of golang installed in final docker image.
- golang code is quite fast in execution time. And this script is being called multiple times each time docker container starts, so it's important for this piece of code to be efficient to ensure we get good enough startup times.
- It can be unit tested and thus gives us confidence in the behaviour of the code.
- It is well tested code which is already implemented.
- Given that it's a very small piece of logic which has a very limited usecase, the maintenance cost associated with it is also quite small.
- We will rarely need to make changes in this logic and when we do, it will be supported by unit tests.

Here are some alternatives for the choice of language, along with their pros and cons:-

**Java**:-
- In case of JVM based Docker image using java to write utility script won't lead to significant image size difference as we already have JRE available to us. Startup time will get impacted though, as java is significantly slower than golang and we are calling this script multiple times.
- For native-image based Docker Image JRE won't be available to us in the final image, so we will need to generate a native executable for the utility scripts. We may need to generate native configs for it as well.
- Given that native-image is still experimental, we would like to avoid using it in JVM based Docker image for the utility script, as this Docker Image can be used for production environment as well.
- Unit tests can be added easily in java, so we will have confidence in the behaviour of the scripts.

**Bash script**:-
- Unit testing the script won't be possible.
- Which means debugging and maintaining it will be more challenging.

**Python**:-
- It will force us to have a python environment to run the utility, increasing the final image size.
- It will be slower than golang or bash.
- It will be unit testable.
- Not a good option as the image size impact and startup time impact are not worth it.

mimaison commented:
Thanks @VedarthConfluent. Once you've updated the KIP, please also ping the DISCUSS or VOTE threads, mentioning the updates so it's visible to all.
